### ISSUE TITLE
load_templates failing to load templates, despite reporting success

### ISSUE THREAD

**Issue by phil-hands** [os-autoinst/openQA#4394]
I think this may well be an issue that is specific to the current state of the database in my server, rather than a general issue, so this is more a request for help diagnosing what's wrong than an attempt to imply general breakage.

Having said that, the most obvious symptom I'm seeing is that I can run the load_templates script, and be given a result listing the numbers of updates that looks like everything worked, but if I then do a dump_templates or look at the webui  it's empty, so that's something that needs fixing somewhere.

I generally use a [slightly tweaked version](https://salsa.debian.org/qa/openqa/openqa-tests-debian/-/blob/debian/fifloader.py) of Fedora's `fifloader.py` for the uploads, which generates the yaml for me, but I've produced identical behaviour using the load_templates script and the [generated.json](https://salsa.debian.org/qa/openqa/openqa-tests-debian/-/blob/debian/generated.json) file produced by `fifloader.py`

The problem appeared when I tried to upload a version of the configuration that was the same as a version that had previously worked, so I don't think the contents of the configuration is the issue (I was attempting to apply an update that it turns out I'd not actually committed to git ;-) )

It's possible that I updated to a more recent version of openqa between the last successful load of configuration and the failure, and I've since upgraded to the latest version from last night (4.6.1639057582.07f0f29) just to make sure it wasn't a temporary blip.

I have a feeling that the transition to YAML for job templates is somehow related to this, as I'm pretty sure that I never saved a new yaml config for any of my job groups, but some of them were behaving as though I had (i.e. not complaining at me) after the first failed upload and possibly since the upgrade to 4.6.1632930736.2895850 that had preceded that, although I didn't notice that.

For now I got back in business by cut&pasting the config via the webUI from an instance on my laptop, which appears to be unaffected -- that was a bit of a pain, so I will make a point of copying the database before experimenting further.

Any help diagnosing what's going on here would be much appreciated.

Additionally, if there are alternative ways of keeping one's whole config in git, and applying it, I'm very happy to explore them, as I wouldn't be surprised if there's a better way of achieving the same ends.

**Comment by perlpunk** [comment]
Can you elaborate on the results you get?

> but if I then do a dump_templates or look at the webui it's empty

What does "empty" here mean? `dump_templates` returns an empty reponse?
When you look at the webui, what page do you mean?

**Comment by Martchus** [comment]
> Additionally, if there are alternative ways of keeping one's whole config in git, and applying it, I'm very happy to explore them, as I wouldn't be surprised if there's a better way of achieving the same ends.

Not really. We currently only support really support the job templates YAML document but that's only covering job templates. The load_templates/dump_templates scripts were there before and should still work but they are not tested very well and might have some rough edges. The scripts are basically just using the API and should normally print some error messages if something goes wrong. Hard to tell what's wrong in your case without more concrete steps to reproduce. Maybe also the server logs contain useful messages (e.g. some SQL errors).

**Comment by okurz** [comment]
> > Additionally, if there are alternative ways of keeping one's whole config in git, and applying it, I'm very happy to explore them, as I wouldn't be surprised if there's a better way of achieving the same ends.
> 
> Not really. We currently only support really support the job templates YAML document but that's only covering job templates. The load_templates/dump_templates scripts were there before and should still work but they are not tested very well and might have some rough edges.

However this approach is actively used, e.g. for https://openqa.fedoraproject.org/ , and AFAIK that has been the case successfully for many years

**Comment by phil-hands** [comment]
> Can you elaborate on the results you get?

Having just tried to reproduce the problem, I see that load_templates has taken to telling me:
<pre>
Bad Request: Groups Debian (amd64), Debian Edu, Kali, Salsa CI must be updated through the YAML template at /usr/share/openqa/script/load_templates line 109.
</pre>
which is because of the way that I had to get things back up and running, which included saving yaml configs.

So, in order to have anything to show you I'm working round that by hacking out the templates thus:
<pre>
psql openqa -c 'UPDATE job_groups SET template = null where template is not null'
</pre>
which is not the first time I've resorted to something like that, so I'd guess that this may well be the source of the problem, but without doing that I'm definitely not going to be able to use load_templates again on that database.

Having done that, I can then run `fifloader.py`, or `load_templates` against the `generated.yaml`, with the same result, which is that sends a copy of the yaml at stdout, as usual, followed by the result response:

<pre>
{
  JobTemplates => { added => 60, of => 60 },
  Machines     => { added => 5, of => 5 },
  Products     => { added => 5, of => 5 },
  TestSuites   => { added => 29, of => 29 },
}
</pre>
which looks like the normal success result to me.  The first obvious difference being that if I run it again immediately (without the --clean option) it gives me exactly the same result, whereas I'd expect subsequent uploads to look more like:

<pre>
{
  JobTemplates => { added => 0, of => 60 },
  Machines     => { added => 0, of => 5 },
  Products     => { added => 0, of => 5 },
  TestSuites   => { added => 0, of => 29 },
}
</pre>
because nothing would have changed between the two.

> > but if I then do a dump_templates or look at the webui it's empty
> 
> What does "empty" here mean? `dump_templates` returns an empty reponse?

<pre>
{
   "JobTemplates" : [],
   "JobGroups" : [
      {
         "template" : "products: {}\nscenarios: {}\n",
         "group_name" : "Debian (amd64)"
      },
      {
         "template" : "products: {}\nscenarios: {}\n",
         "group_name" : "Debian Edu"
      },
      {
         "group_name" : "Kali",
         "template" : "products: {}\nscenarios: {}\n"
      },
      {
         "group_name" : "Salsa CI",
         "template" : "products: {}\nscenarios: {}\n"
      }
   ],
   "Machines" : [],
   "TestSuites" : [],
   "Products" : []
}
</pre>

> When you look at the webui, what page do you mean?


![2021-12-10-221839_1169x589_scrot](https://user-images.githubusercontent.com/64984191/145644889-d2d4ff34-7c37-42a5-9df8-f5909991afa7.png)

I'll dig into this further tomorrow if I have time, or more likely Monday.

**Comment by phil-hands** [comment]
This issue turns out to be caused by an apache configuration problem.

I had apache redirecting http to https. This all worked as far as I can tell, but the way I'd done it meant that if you connected to `http://localhost` you'd also get redirected to `https://openqa.debian.net`.

`load_templates` connects to `localhost` by default, so when the redirect was added that's presumably when things broke, but I didn't realise because I didn't try any updates for a while.

It only occurred to me today to try specifying the `--host` option, to force it to connect to `openqa.debian.net`, which worked fine.

I've fixed the problem by adding a virtual host for `localhost` to the apache config, also offering the openqa site.

The odd thing about this is that with the old config `load_templates` gives every appearance of working, and it's clearly connecting well enough to clear out the database, but the rest of the upload doesn't land in the database, resulting in a  mostly empty configuarion.

This makes me wonder if there is something that checks that the name the client is trying to connect to matches the one it actually connected to, or some such, and reacts badly if not.

I could just close this at this point, but thought I'd leave it open for a little while in case anyone actually cares that this weird case exists (I suspect not, given that `load_templates` seems to be a thing of the past)
